pkg/controller/node_availability.go

```diff
@@ -0,0 +1,150 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"sync/atomic"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+	azClientSet "sigs.k8s.io/azuredisk-csi-driver/pkg/apis/client/clientset/versioned"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+type ReconcileNodeAvailability struct {
+	client                client.Client
+	namespace             string
+	controllerSharedState *SharedState
+	azVolumeClient        azClientSet.Interface
+	kubeClient            kubernetes.Interface
+}
+
+var _ reconcile.Reconciler = &ReconcileNodeAvailability{}
+
+func (r *ReconcileNodeAvailability) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+
+	n := &corev1.Node{}
+	err := r.client.Get(ctx, request.NamespacedName, n)
+
+	if errors.IsNotFound(err) {
+		return reconcile.Result{}, nil
+	}
+
+	if err == nil {
+		if !n.Spec.Unschedulable {
+			//Node is schedulable, proceed to attempt creation of replica attachment
+			if atomic.SwapInt32(&r.controllerSharedState.processingReplicaRequestQueue, 1) == 0 {
+				for len(r.controllerSharedState.priorityReplicaRequestsQueue.queue.List()) > 0 {
+					replicaRequest := r.controllerSharedState.priorityReplicaRequestsQueue.Pop()
```

Comment:
It's unfortunate that `cache.Heap.Pop` is blocking thus forcing you to `List` in order to see if the queue is empty. This can be a relatively expensive operation. Consider tracking the count of items in the queue separately in `VolumeReplicaRequestsPriorityQueue`. You can use [`atomic.AddInt32`](https://pkg.go.dev/sync/atomic#AddInt32) to update the count in the `Push` and `Pop` methods.

Reply:
I'm also wondering if, with my implementation, the operation in addToOperationQueue completes before this loop is finished, and the attachment fails and volume is added back to the queue, this would be an infinite loop? or, if it requeues the request and the priority is still higher than the other volumes, this would also never reach those lower ranked volumes. Any thoughts? I could store volumes already tried and check against that before attempting addToOperationQueue, not sure if that would be expensive

Reply:
Yeah, that's a good point and we'll need to defend against this scenario. I don't really like the idea of an exclusion list, but I don't have any ideas that still allow us to use `client-go/tools/cache/Heap`. If you go back to using our own priority queue implementation, we could implement a `func Drain() []*ReplicaRequest` function that would atomically empty and return all elements on the queue as an array in priory order. This would allow concurrent and requeue requests to be made w/o resulting in an infinite loop.